vignettes/custom-functions.Rmd

```diff
@@ -0,0 +1,109 @@
+---
+title: "Using custom functions or other R packages"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Using custom functions or other R packages}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r, include = FALSE}
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+```{r setup}
+library(neopolars)
+```
+
+`polars` contains a large set of functions to manipulate variables, whether they are numerical, strings, dates, or other types.
+Still, it is sometimes necessary to apply custom functions either written by you or available in other R packages.
+This vignette details several ways to do that in the most computationally efficient way.
+
+## Writing functions using `polars` expressions
+
+One of `polars`' biggest strength is the composability of expressions.
+Chaining multiple functions to create custom functions is quite simple.
+For example, `polars` doesn't provide a built-in function to standardize a variable, but we could create one using expressions:
+
+```{r}
+pl_standardize <- function(x) {
+  (x - x$mean()) / x$std()
+}
+```
+
+And we can then run it in `$with_columns()` for instance:
+```{r}
+dat <- as_polars_df(mtcars[, c("carb", "mpg", "drat")])
+dat$with_columns(
+  carb_stand = pl_standardize(pl$col("carb"))
+)
+```
+
+## Using `purrr`
+
+`polars` expressions allow one to write their own functions, but it is sometimes necessary to use functions that already exist in other R packages.
+When this is the case, one should keep in mind that applying functions that are external to `polars` will limit `polars`' strengths.
+In particular, `polars` is fast in part because it is able to detect the various transformations applied to the data and run them in parallel or cache them so that only the strict minimum of computations is made.
+Using `polars` expressions will also allow the code to be run in streaming mode, allowing for larger-than-memory data.
+When one uses functions that `polars` doesn't natively know, those benefits are lost.
+This is the reason why it is better to use `polars` expressions in your own functions if you can.
+
+To apply an external function to a `polars` object, we must convert this object to R, apply the function, and then convert it back to `polars`.
+If you come from Python Polars, this is a big difference.
+In Python Polars, several functions (for instance `map_batches()` and `map_elements()`) allow one to pass custom functions.
+Those are not in the R implementation because we cannot guarantee that they will function correctly in streaming mode.
```

Comment:
I think this is worth mentioning, but I'm not sure we should explain why it's not in the R implementation (maybe I also missed some reasons for why it is absent).